src/pages/api/auth/[...nextauth].ts

```diff
@@ -12,26 +17,53 @@ export default NextAuth({
 		 *
 		 * ref: https://next-auth.js.org/configuration/callbacks#jwt-callback
 		 */
-		async jwt({ token, user, account, profile }) {
-			if (user && account) {
-				token['https://auth.hashicorp.com/auth0/connection/id'] =
-					user['https://auth.hashicorp.com/auth0/connection/id']
+		async jwt({ token, account, profile }) {
+			const isInitial = !!account && !!profile
+			isDev &&
+				console.log('jwt callback (%s)', isInitial ? 'initial' : 'subsequent')
+
+			// initial call during sign in
+			if (isInitial) {
+				// persist access_token and refresh_token on the session token
+				token.access_token = account.access_token
+				token.refresh_token = account.refresh_token
 
-				token.cloud_idp_access_token = account.access_token
+				// Picture is passed to session.user.image
 				token.picture = profile.picture as string
 				token.nickname = profile.nickname
 
-				token.exp = user.exp
-			}
-
-			if (Date.now() / 1000 < (token as { exp: number }).exp) {
 				return token
 			}
 
-			// TODO: validate that the refresh token flow works as implemented
-			// return refreshAccessToken(token)
+			// subsequent calls when session is accessed
+			const [isAccessTokenExpired, secondsUntilExpiry] = isJwtExpired(
+				token.access_token as string
+			)
+
+			if (isAccessTokenExpired) {
+				// Refresh token set w/ refreshToken
+				try {
+					isDev && console.log('access token has expired; refreshing...')
+					const { access_token, refresh_token } = await refreshTokenSet(
+						token.refresh_token as string
+					)
+					isDev && console.log('successfully refreshed token set')
+
+					token.access_token = access_token
+					token.refresh_token = refresh_token
+				} catch (err) {
+					return { ...token, error: AuthErrors.RefreshAccessTokenError }
+				}
```

Comment:
💭 Would it be good to `console.error` here as well? I'm not sure if datadog catches those logs yet or not

Reply:
I'm on board!

Reply:
We should have Datadog log drains already set up — see https://vercel.com/hashicorp/dev-portal/settings/integrations